### PR TITLE
[SELS_TASK] [FE] Lesson (Category) Answering Page

### PR DESCRIPTION
**Pull request by ww-jermaine**
**[Asana Link]**
https://app.asana.com/0/1202968674806069/1202968674806110/f

**[Commands]**
Install the needed packages listed in the `package.json`.

```
cd frontend
npm install
```

Run the command for development build.
```
npm start
```

Setup the virtual environment
```
cd backend
venv\Scripts\Activate
```

Run the application
```
python manage.py runserver
```

**[Expected Output]** and **[Screenshots]**

image

image

image

image

image

### CODE REVIEW

File: backend/api/urls.py [hunk 1]
```diff
@@ -5,7 +5,12 @@
 from rest_framework import permissions
 from rest_framework.routers import SimpleRouter
 
-from elearning.views import CategoryViewSet, WordViewSet
+from elearning.views import (
+    CategoryViewSet,
+    LessonResultExistsView,
```
Comment (Alvilb) [inline backend/api/urls.py]: unused import?

File: backend/elearning/serializers.py [hunk 2]
```diff
@@ -117,6 +148,32 @@ class Meta:
...
+    category_taken = serializers.CharField(source="category_taken.category_name")
+    user_profile_taker = serializers.CharField(source="user_profile_taker.user.username")
+
+    class Meta:
+        model = QuizRecord
+        fields = ("id", "user_profile_taker", "category_taken", "score", "total", "words_learned")
+
+    def get_words_learned(self, obj: QuizRecord) -> List[WordRecord]:
+        qs = WordRecord.objects.filter(quiz_record=obj)
+        serializer = WordRecordSerializer(qs, many=True)
```
Comment (Alvilb) [inline backend/elearning/serializers.py]: use a more descriptive variable name than `qs`

File: backend/elearning/views.py [hunk 3]
```diff
@@ -109,14 +110,51 @@ def LessonAnsweringPostView(taker_id: int, lesson: LessonAnsweringSchema):
...
+    http_method_names = ["get"]
+    my_tags = ["Lesson Results"]
+
+    # lookup_field = "user_profile_taker_id"
+
+    # def retrieve(self, request, *args, **kwargs):
+    #    user_profile_taker_id = kwargs["user_profile_taker_id"]
+    #    user_profile = UserProfile.objects.get(id=user_profile_taker_id)
+    #    self.queryset = QuizRecord.objects.filter(user_profile_taker=user_profile)
+    #    return super(LessonResultsViewSet, self).retrieve(request, *args, **kwargs)
```
Comment (Alvilb) [inline backend/elearning/views.py]: will these be used? please add a TODO note if yes

File: backend/elearning/serializers.py [hunk 4]
```diff
@@ -117,6 +148,32 @@ class Meta:
...
+        return serializer.data
+        # return default_serializers.serialize("json", qs)
```
Comment (Alvilb) [inline backend/elearning/serializers.py]: will we need this?

File: backend/elearning/serializers.py [hunk 5]
```diff
@@ -26,19 +28,47 @@ class Meta:
...
         model = Answer
-        fields = ("word_id", "answer_text")
-        read_only_fields = ("word_id",)
-        # depth = 1
+        fields = ("answer_text",)
+
+
+class WordCatSerializer(serializers.ModelSerializer):
+    choices = ChoiceSerializer(many=True, read_only=True)
+    # answer = AnswerSerializer(read_only=True)
```
Comment (Alvilb) [inline backend/elearning/serializers.py]: here are as well do we need this?

File: frontend/src/App.js [hunk 6]
```diff
@@ -32,6 +24,7 @@ function App() {
     )
   }
 
+  //const [user, setUser] = useState(null);
```
Comment (Alvilb) [inline frontend/src/App.js]: please add note why we it is left commented out here

File: frontend/src/hooks/useGetLesson.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+export default function useGetLesson(category_id) {
+  const [lesson, setLesson] = useState({})
+  const [isLoading, setIsLoading] = useState(true)
+  const [error, setError] = useState({})
+
+  useEffect(() => {
+    if (category_id !== undefined) {
+      categoryService
+        .getLesson(category_id)
+        .then((response) => {
```
Comment (Alvilb) [inline frontend/src/hooks/useGetLesson.js:13]: no need to use states here, just use variables

```
let lesson = {}
let isLoading = true
let error = {}

useEffect(() => {
...
```

File: frontend/src/hooks/useGetLessons.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+export default function useGetLessons() {
+  const [lessons, setLessons] = useState([])
+  const [isLoading, setIsLoading] = useState(true)
+  const [error, setError] = useState({})
+
+  useEffect(() => {
+    categoryService
+      .getCategories()
+      .then((response) => {
```
Comment (Alvilb) [inline frontend/src/hooks/useGetLessons.js:12]: here as well no need for useState

File: frontend/src/hooks/useLessonExists.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import categoryService from '../services/category.service'
+
+export default function useLessonExists(
+  category_taken_id,
+  user_profile_taker_id,
+) {
+  const [exists, setExists] = useState(false)
+  const [isLoading, setIsLoading] = useState(true)
+  const [error, setError] = useState({})
+
```
Comment (Alvilb) [inline frontend/src/hooks/useLessonExists.js]: here as well just use variables

File: frontend/src/hooks/useLessonResults.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+export default function useLessonResults(
+  category_taken_id,
+  user_profile_taker_id,
+) {
+  const [resultData, setResultData] = useState({})
+  const [isLoading, setIsLoading] = useState(true)
+  const [error, setError] = useState({})
+
+  useEffect(() => {
+    if (
```
Comment (Alvilb) [inline frontend/src/hooks/useLessonResults.js]: no need for useState

File: frontend/src/pages/category-page/CategoryPage.js [hunk 11]
```diff
@@ -0,0 +1,259 @@
+import { useParams, Navigate } from 'react-router-dom'
+import Footer from '../../components/Footer'
+import NavBarLanding from '../../components/elements/NavBarLanding'
+import useLessonExists from '../../hooks/useLessonExists'
+import useGetLesson from '../../hooks/useGetLesson'
+import CategoryService from '../../services/category.service'
+import { useState } from 'react'
+
+export default function CategoryPage() {
```
Comment (Alvilb) [inline frontend/src/pages/category-page/CategoryPage.js:11]: break down into smaller components
DisplayQuestion
ChoiceItem
StartButton
QuizDescription

File: frontend/src/pages/category-results/CategoryResultsPage.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
+import Footer from '../../components/Footer'
+import NavBarLanding from '../../components/elements/NavBarLanding'
+//import resultData from './results.json'
```
Comment (Alvilb) [inline frontend/src/pages/category-results/CategoryResultsPage.js]: remove commented out codes

File: frontend/src/pages/category-results/CategoryResultsPage.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+import { Row, Col, Container } from 'react-bootstrap'
+import useLessonResults from '../../hooks/useLessonResults'
+import { useParams, Navigate } from 'react-router-dom'
+
+export default function CategoryResultsPage() {
```
Comment (Alvilb) [inline frontend/src/pages/category-results/CategoryResultsPage.js]: breakdown into small components
Load resultData
GridSystem
Item

File: frontend/src/hooks/useLessonExists.js (hunk 9 above)
Comment (marcjermaine-pontiveros) [reply frontend/src/hooks/useLessonExists.js:11]: image

File: frontend/src/hooks/useGetLesson.js (hunk 7 above)
Comment (ww-jermaine) [reply frontend/src/hooks/useGetLesson.js:13]: image

File: frontend/src/hooks/useLessonExists.js (hunk 9 above)
Comment (ww-jermaine) [reply frontend/src/hooks/useLessonExists.js]: image

File: backend/elearning/views.py (hunk 3 above)
Comment (ww-jermaine) [reply backend/elearning/views.py]: Placed a todo note sir, thank you!

File: frontend/src/hooks/useLessonExists.js (hunk 9 above)
Comment (Alvilb) [reply frontend/src/hooks/useLessonExists.js]: lets use your original changes `useState`

**Review by Alvilb — APPROVED** [review]
Approved

**Pull request merged**